runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java

```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
+        ExternalTransforms.ExternalConfigurationPayload.newBuilder()
+            .setSchema(SchemaTranslation.schemaToProto(configRow.getSchema(), true))
+            .setPayload(outputStream.toByteString())
+            .build();
+
+    String serviceAddress = null;
+    TransformServiceLauncher service = null;
+    try {
+      if (externalTranslationOptions.getTransformServiceAddress() != null) {
+        serviceAddress = externalTranslationOptions.getTransformServiceAddress();
+      } else if (externalTranslationOptions.getTransformServiceBeamVersion() != null) {
+        String projectName = UUID.randomUUID().toString();
+        service = TransformServiceLauncher.forProject(projectName, findAvailablePort());
+        service.setBeamVersion(externalTranslationOptions.getTransformServiceBeamVersion());
+
+        // Starting the transform service.
+        service.start();
+        // Waiting the service to be ready.
+        service.waitTillUp(15000);
+      } else {
+        throw new IllegalArgumentException(
+            "Either option TransformServiceAddress or option TransformServiceBeamVersion should be provided to override a transform using the transform service");
+      }
+
+      if (serviceAddress == null) {
+        throw new IllegalArgumentException(
+            "Cannot override the transform "
+                + urn
+                + " since a valid transform service address could not be determined");
+      }
+
+      // Creating an ExternalTransform and expanding it using the transform service.
+      // Input will be the same input provided to the transform bing overridden.
+      ExpandableTransform<InputT, OutputT> externalTransform =
+          (ExpandableTransform<InputT, OutputT>)
+              External.of(urn, payload.toByteArray(), serviceAddress);
+
+      PCollectionTuple input = PCollectionTuple.empty(pipeline);
+      for (TupleTag<?> tag : (Set<TupleTag<?>>) appliedPTransform.getInputs().keySet()) {
+        PCollection<?> pc = appliedPTransform.getInputs().get(tag);
+        if (pc == null) {
+          throw new IllegalArgumentException(
+              "Input of transform " + appliedPTransform + " with tag " + tag + " was null.");
+        }
+        input = input.and(tag, (PCollection) pc);
+      }
+      POutput output = externalTransform.expand((InputT) input);
+
+      // Outputs of the transform being overridden.
+      Map<TupleTag<?>, PCollection<?>> originalOutputs = appliedPTransform.getOutputs();
+
+      // After expansion some transforms might still refer to the output of the already overridden
+      // transform as their input.
+      // Such inputs have to be overridden to use the output of the new upgraded transform.
+      Map<String, String> inputReplacements = new HashMap<>();
+
+      // Will contain the outputs of the upgraded transform.
+      Map<TupleTag<?>, PCollection<?>> newOutputs = new HashMap<>();
+
+      if (output instanceof PCollectionTuple) {
```

Comment:
This code deals with both the Pipeline object and the Pipeline proto, and it's a bit hard to follow where the source of truth is. My personal preference would be to simply act on the protos consuming the expansion service response directly. Honestly, I don't think that'd be any more complicated than what you're already having to do here already.